### PR TITLE
SpaceSaver fromBytes & toBytes (#577)

### PR DESCRIPTION
**Pull request by ElPicador**
My implementation is very basic, but it should work

### CODE REVIEW

File: algebird-test/src/test/scala/com/twitter/algebird/SpaceSaverTest.scala [hunk 1]
```diff
@@ -59,5 +61,26 @@ class SpaceSaverTest extends WordSpec with Matchers {
     "be consistent in its constructors" in {
       assert(SpaceSaver(10, "ha") ++ SpaceSaver(10, "ha") ++ SpaceSaver(10, "ha") === SpaceSaver(10, "ha", 3))
     }
+
+    "Correctly serialize" in {
+      //SSOne
+      verifySerialization(SpaceSaver(40, "string that is not empty"))
```
Comment (sritchie-stripe) [inline algebird-test/src/test/scala/com/twitter/algebird/SpaceSaverTest.scala]: would you mind converting this to a scalacheck test? Something like

```scala
property("SpaceSaver serializes properly") { ss: SpaceSaver[String] =>
  val spaceSaverAsArrayByte = SpaceSaver.toBytes(ss, stringToArrayByte)
  SpaceSaver.fromBytes(spaceSaverAsArrayByte, arrayByteToString).get == ss
}
```

The best way to make this work is to get a scalacheck `Arbitrary` instance into this file:

https://github.com/twitter/algebird/blob/develop/algebird-test/src/main/scala/com/twitter/algebird/scalacheck/Arbitrary.scala

and a `Gen` instance into this file:

https://github.com/twitter/algebird/blob/develop/algebird-test/src/main/scala/com/twitter/algebird/scalacheck/Gen.scala

I know it's a bigger ask, but serialization is a big deal and I think we absolutely need the scalacheck laws in place here.

File: algebird-core/src/main/scala/com/twitter/algebird/SpaceSaver.scala [hunk 2]
```diff
@@ -18,6 +21,98 @@ object SpaceSaver {
...
 
   implicit def spaceSaverSemiGroup[T]: Semigroup[SpaceSaver[T]] = new SpaceSaverSemigroup[T]
+
+  /**
+   * Encodes the SpaceSaver as a sequence of bytes containing in order
+   * - 1 byte: 1/2 => 1 = SSOne, 2 = SSMany
+   * - 4 bytes: the capacity
+   * - N bytes: the item/counters (counters as length + N*(item size + item + 2 * counters)
+   */
+  def toBytes[T](ss: SpaceSaver[T], tSerializer: T => Array[Byte]): Array[Byte] = {
```
Comment (sritchie-stripe) [inline algebird-core/src/main/scala/com/twitter/algebird/SpaceSaver.scala:31]: Another ask - could we serialize a couple of examples and put a base64-encoded string of the bytes into the test, so that we can guarantee that the serialization doesn't change? I can see us making a change here that wouldn't break the tests, but would break deserialization of old instances.

**Review by sritchie-stripe — CHANGES_REQUESTED** [review]
Thanks for doing this, and love the coverage increase! A few testing asks, since serialization is a big deal. Really happy that we're getting this done.

**Comment by sritchie-stripe** [comment]
@ElPicador @non @johnynek have any of you used scodec? https://github.com/scodec/scodec Seems like it might be a nice library to use for these serialization as we start writing them.

**Comment by sritchie-stripe** [comment]
Oh, @ElPicador, one more ask - would you please add an example of the new serialization methods to the spacesaver documentation page?

https://github.com/twitter/algebird/blob/develop/docs/src/main/tut/datatypes/approx/space_saver.md

Just something about like this 

```
    ## Serialization
   You can serialize `SpaceSaver` instances using `fromBytes` and `toBytes`:

    ```tut:book
    import com.twitter.algebird._
    val ss = ???
    // keep the example going!
    ```
```
With the new documentation, you can just add lines as you'd type them into the repl. The output will look great:

https://twitter.github.io/algebird/datatypes/averaged_value.html

Thanks again!

File: algebird-test/src/test/scala/com/twitter/algebird/SpaceSaverTest.scala (hunk 1 above)
Comment (ElPicador) [reply algebird-test/src/test/scala/com/twitter/algebird/SpaceSaverTest.scala]: Will do. I wanted to do it, but before I wanted to be sure the PR could be accepted 😄

File: algebird-core/src/main/scala/com/twitter/algebird/SpaceSaver.scala (hunk 2 above)
Comment (ElPicador) [reply algebird-core/src/main/scala/com/twitter/algebird/SpaceSaver.scala:31]: Good idea, will do

**Comment by ElPicador** [comment]
It should be good

File: algebird-core/src/main/scala/com/twitter/algebird/SpaceSaver.scala [hunk 3]
```diff
@@ -18,6 +21,98 @@ object SpaceSaver {
...
+        buffer.result.toArray
+    }
+  }
+
+  // Make sure to be reversible so fromBytes(toBytes(x)) == x
+  def fromBytes[T](bytes: Array[Byte], tDeserializer: Array[Byte] => Try[T]): Try[SpaceSaver[T]] = {
+    fromByteBuffer(ByteBuffer.wrap(bytes), tDeserializer)
+  }
+
+  def fromByteBuffer[T](bb: ByteBuffer, tDeserializer: Array[Byte] => Try[T]): Try[SpaceSaver[T]] = {
```
Comment (johnynek) [inline algebird-core/src/main/scala/com/twitter/algebird/SpaceSaver.scala]: I think `ByteBuffer => Try[T]` might allow some better performance at a moment.

Comment (ElPicador) [reply algebird-core/src/main/scala/com/twitter/algebird/SpaceSaver.scala]: Changed

File: algebird-core/src/main/scala/com/twitter/algebird/SpaceSaver.scala [hunk 4]
```diff
@@ -18,6 +21,98 @@ object SpaceSaver {
...
+    }
+  }
+
+  // Make sure to be reversible so fromBytes(toBytes(x)) == x
+  def fromBytes[T](bytes: Array[Byte], tDeserializer: Array[Byte] => Try[T]): Try[SpaceSaver[T]] = {
+    fromByteBuffer(ByteBuffer.wrap(bytes), buffer => tDeserializer(buffer.array()))
+  }
+
+  def fromByteBuffer[T](bb: ByteBuffer, tDeserializer: ByteBuffer => Try[T]): Try[SpaceSaver[T]] = {
+    bb.get.toInt match {
```
Comment (johnynek) [inline algebird-core/src/main/scala/com/twitter/algebird/SpaceSaver.scala]: Couldn't other operations on `bb` throw? I think we need an outer Try to catch those exceptions too (or use a try/catch block inside the method.

File: algebird-core/src/main/scala/com/twitter/algebird/SpaceSaver.scala [hunk 5]
```diff
@@ -18,6 +21,98 @@ object SpaceSaver {
...
+            case Failure(e) =>
+              return Failure(e)
+            case Success(i) =>
+              counters += ((i, (a, b)))
+          }
+
+          countersToDeserialize -= 1
+        }
+
+        Try(SSMany(capacity, counters.toMap))
```
Comment (johnynek) [inline algebird-core/src/main/scala/com/twitter/algebird/SpaceSaver.scala]: Why do you need Try here rather than just Success? Could this throw?

File: algebird-core/src/main/scala/com/twitter/algebird/SpaceSaver.scala [hunk 6]
```diff
@@ -18,6 +21,98 @@ object SpaceSaver {
...
+          val itemLength = bb.getInt()
+          val itemAsBytes = new Array[Byte](itemLength)
+          bb.get(itemAsBytes)
+          val item = tDeserializer(ByteBuffer.wrap(itemAsBytes))
+
+          val a = bb.getLong
+          val b = bb.getLong
+
+          item match {
+            case Failure(e) =>
```
Comment (johnynek) [inline algebird-core/src/main/scala/com/twitter/algebird/SpaceSaver.scala]: `case f@Failure(_) => return f` would avoid an allocation.

File: algebird-test/src/test/scala/com/twitter/algebird/SpaceSaverTest.scala [hunk 7]
```diff
@@ -26,22 +31,31 @@ class SpaceSaverLaws extends CheckProperties {
...
             (left consistentWith right) && (right consistentWith left)
           }
+
         commutativeSemigroupLaws[SpaceSaver[Int]]
       }
     }
   }
+
+  property("SpaceSaver can serialize/deserialize itself") {
+    forAll { ss: SpaceSaver[String] =>
```
Comment (johnynek) [inline algebird-test/src/test/scala/com/twitter/algebird/SpaceSaverTest.scala:41]: Can we also add a test that exercises Failure? Like pass an empty array, and see that you get Failure not an exception thrown.

So basically: `forAll { a: Array[Byte] => ` deserialization does not throw? You get Success or Failure.

**Review by johnynek — CHANGES_REQUESTED** [review]

File: algebird-test/src/test/scala/com/twitter/algebird/SpaceSaverTest.scala (hunk 7 above)
Comment (ElPicador) [reply algebird-test/src/test/scala/com/twitter/algebird/SpaceSaverTest.scala:41]: Done

File: algebird-core/src/main/scala/com/twitter/algebird/SpaceSaver.scala (hunk 5 above)
Comment (ElPicador) [reply algebird-core/src/main/scala/com/twitter/algebird/SpaceSaver.scala]: True, I changed it

File: algebird-core/src/main/scala/com/twitter/algebird/SpaceSaver.scala (hunk 6 above)
Comment (ElPicador) [reply algebird-core/src/main/scala/com/twitter/algebird/SpaceSaver.scala]: Won't compile. One is a `Try[T]` the other is a `Try[SpaceSaver[T]]`

File: algebird-core/src/main/scala/com/twitter/algebird/SpaceSaver.scala (hunk 4 above)
Comment (ElPicador) [reply algebird-core/src/main/scala/com/twitter/algebird/SpaceSaver.scala]: You're right. I've changed it

File: algebird-test/src/test/scala/com/twitter/algebird/SpaceSaverTest.scala (hunk 7 above)
Comment (sritchie) [reply algebird-test/src/test/scala/com/twitter/algebird/SpaceSaverTest.scala:41]: Pending this one change this looks great to me! @ElPicador, thanks for responding to these rounds of feedback. With this extra test the new serde code will have 100% coverage :)

Comment (ElPicador) [reply algebird-test/src/test/scala/com/twitter/algebird/SpaceSaverTest.scala:41]: @sritchie I'm not sure to understand what change you are talking about.

Comment (sritchie-stripe) [reply algebird-test/src/test/scala/com/twitter/algebird/SpaceSaverTest.scala:41]: oh, sorry, I was thinking of the test that @johnynek requested.

**Comment by sritchie-stripe** [comment]
looks great to me! @ElPicador, @johnynek is traveling today - let's let him take one more look, and then we'll get this merged.

The plan is to fix up the remaining two issues in #600 and then prep a release of 0.13.0. That will be binary incompatible in a couple of minor ways that should be easy to fix. Does that work for you?

Thanks so much for the contribution!

**Comment by ElPicador** [comment]
Perfect for me. My step is to add a serializer in twitter-chill.

**Review by johnynek — APPROVED** [review]

**Comment by johnynek** [comment]
Thanks!
